BZ1844942 - Amending procedure to define explicit resource quotas

pneedle-rh opened the pull request:
Applies to 4.2 onward.

This relates to https://bugzilla.redhat.com/show_bug.cgi?id=1844942.

A preview is available at https://bz1844942--ocpdocs.netlify.app/openshift-enterprise/latest/applications/quotas/quotas-setting-per-project.html#configuring-explicit-resource-quotas_quotas-setting-per-project.

pneedle-rh commented:
@openshift/team-documentation for peer review please. Please can someone also add 'Peer Review Needed', 'branch/enterprise-4.2', 'branch/enterprise-4.3', 'branch/enterprise-4.4' labels and 'Next Release' milestone? Thanks!

pneedle-rh commented:
@xltian can you please review this from a QE perspective? Thanks!

kalexand-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -8,44 +8,100 @@
+. Reference the new project request template in the cluster's project configuration resource:
+.. Access the project configuration resource for editing:
++
+** Using the web console:
```
> ! using n
To avoid ambiguity, replace this gerund with either “by using” or “that uses.”

pneedle-rh replied:
Resolved.

kalexand-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -8,44 +8,100 @@
+----
+
+. Verify that the resource quota is applied when new projects are created:
+.. Create a new project:
```
~new~

pneedle-rh replied:
Resolved.

kalexand-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -8,44 +8,100 @@
+.. Create a new project:
++
+----
+oc new-project my-project
```
$ oc

I'd make the project name a variable instead of a discrete value.

pneedle-rh replied:
Resolved.

kalexand-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -8,44 +8,100 @@
+.. Verify that the resource quota has been applied. The following example describes the `storage-consumption` resource quota:
++
+----
+oc describe resourcequotas storage-consumption
```
$ oc

Same as the last command, I'd make the resource quota a variable and explain how you get the name.

pneedle-rh replied:
Resolved.

pneedle-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -8,44 +8,100 @@
-project was not explicitly given quota for storage related to the gold storage
-class, users of that project would not be able to create any storage of that
-type.
+In {product-title} {product-version} it is not possible to configure the `ResourceQuota` admission plug-in. Instead, explicit resource quotas can be defined for a specific resource in the default project template. Those resource quotas are then applied to newly created projects.
```
Thank you for your review! I have subsequently removed most of the concept information that was in the first two paragraphs. I think the single sentence that you suggested is suffice.

pneedle-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -8,44 +8,100 @@
 = Requiring explicit quota to consume a resource
```
I have renamed the title and the ID accordingly. I can't find any reference to the previous ID in the docs.

pneedle-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -8,44 +8,100 @@
 = Requiring explicit quota to consume a resource
 
 If a resource is not managed by quota, a user has no restriction on the amount
```
Thank you. I have reduced the procedural intro down to the one sentence you suggested. On reflection I think nothing else is required as the assembly introduction provides an adequate description about when quotas are needed.

pneedle-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -8,44 +8,100 @@
-type.
+In {product-title} {product-version} it is not possible to configure the `ResourceQuota` admission plug-in. Instead, explicit resource quotas can be defined for a specific resource in the default project template. Those resource quotas are then applied to newly created projects.
+
+[NOTE]
```
I have now removed this and replaced it with the single sentence suggested in your earlier comment.

pneedle-rh replied:
@kalexand-rh reading this comment again, I am unsure if this is in reference to the note or the paragraph preceding it. Can you please confirm?

pneedle-rh replied:
On reflection I think this comment relates to the note, given the position of the comment. I included that note because the Bugzilla mentions that this functionality is possible through webhook admission plug-ins. On review, I agree it is confusing. Plus, it doesn't really add anything without a procedure showing the reader how to do this. That procedure would be very detailed and technical. As such, I will remove the note so that the document can focus on providing the simpler method.

pneedle-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -8,44 +8,100 @@
+
+* Access to the cluster as a user with the cluster-admin role.
+
+* The OpenShift Container Platform CLI (oc) installed.
```
Resolved.

bergerhoffer replied:
Just happened to see this fly by in email. Not critical, but I think we said that this was overkill. We've been able to condense this to below, we just haven't updated all existing instances that were worded this way.

```
Install the OpenShift CLI (`oc`)
```

pneedle-rh replied:
Thank you Andrea. Based on your comment, I have now updated the prerequisite to 'Install the OpenShift CLI (`oc`)'.

pneedle-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -8,44 +8,100 @@
+<3> Across all persistent volume claims in a project, the sum of storage requested in the gold storage class cannot exceed this value.
+<4> Across all persistent volume claims in a project, the sum of storage requested in the silver storage class cannot exceed this value.
+<5> Across all persistent volume claims in a project, the total number of claims in the silver storage class cannot exceed this value.
+<6> Across all persistent volume claims in a project, the sum of storage requested in the bronze storage class cannot exceed this value. When this is set to 0, it means bronze storage class cannot request storage.
```
Resolved.

pneedle-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -8,44 +8,100 @@
+<6> Across all persistent volume claims in a project, the sum of storage requested in the bronze storage class cannot exceed this value. When this is set to 0, it means bronze storage class cannot request storage.
+<7> Across all persistent volume claims in a project, the sum of storage requested in the bronze storage class cannot exceed this value. When this is set to 0, it means bronze storage class cannot create claims.
+
+. Create a new project request template from the modified `template.yaml` file, in the `openshift-config` namespace. By default, the template will be called `project-request`:
```
Resolved.

pneedle-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -8,44 +8,100 @@
+<4> Across all persistent volume claims in a project, the sum of storage requested in the silver storage class cannot exceed this value.
+<5> Across all persistent volume claims in a project, the total number of claims in the silver storage class cannot exceed this value.
+<6> Across all persistent volume claims in a project, the sum of storage requested in the bronze storage class cannot exceed this value. When this is set to 0, it means bronze storage class cannot request storage.
+<7> Across all persistent volume claims in a project, the sum of storage requested in the bronze storage class cannot exceed this value. When this is set to 0, it means bronze storage class cannot create claims.
```
Resolved.

xingxingxia commented:
Checked the PR, it uses content of https://docs.openshift.com/container-platform/4.1/applications/projects/configuring-project-creation.html for setting the storageclass quota for project. Node subteam has cases OCP-12827 OCP-11056 OCP-11685 about storageclass quota and will review the part of this.

weinliu commented:
/LGTM
Checked on 
```
$ oc version
Client Version: 4.4.0-rc.10
Server Version: 4.4.0-0.nightly-2020-06-15-203705
Kubernetes Version: v1.17.1+912792b

```
The new steps in the doc is working

```
$ oc describe resourcequotas 
Name:                                                      storage-consumption
Namespace:                                                 weinliu3
Resource                                                   Used  Hard
--------                                                   ----  ----
bronze.storageclass.storage.k8s.io/persistentvolumeclaims  0     0
bronze.storageclass.storage.k8s.io/requests.storage        0     0
gold.storageclass.storage.k8s.io/requests.storage          0     10Gi
persistentvolumeclaims                                     0     10
requests.storage                                           0     50Gi
silver.storageclass.storage.k8s.io/persistentvolumeclaims  0     5
silver.storageclass.storage.k8s.io/requests.storage        0     20Gi
```

pneedle-rh commented:
@kalexand-rh and @weinliu, I have just pushed further changes to this PR, in relation to Bugzilla comment https://bugzilla.redhat.com/show_bug.cgi?id=1844942#c9.

If `oc apply` is used to update an existing project request template, as suggested in BZ comment, the problem is, we get into a situation where the reader needs to understand declarative and imperative methods for updating configuration, which is a bit like pulling at a thread within the procedure. It depends on which method they use in their cluster.

Instead I have updated the procedure to use the following alternative. Frederic Giloux (the Bugzilla reporter) is happy with this solution and it worked in my testing:

~~~
$ oc get templates -n openshift-config
$ oc edit template project-request -n openshift-config
~~~

I also replaced reference to 'default project template' in the module with 'project request template' for consistency. Plus, I have now included the term 'bootstrap project template' for the initial template output to the YAML file, to differentiate it from the project request template.

Please can you let both me know if you are happy with the updated doc? Thanks!

kalexand-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -4,48 +4,131 @@
+
+* Access to the cluster as a user with the cluster-admin role.
+
+* Install the OpenShift CLI (`oc`).
```
```
Install the OpenShift Command-line Interface (CLI), commonly known as `oc`.
```
This is the language that accepted in the [contribution guide](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/term_glossary.adoc#openshift-cli).

pneedle-rh replied:
Thank you for your further review. I had seen your previous suggestion to change this, but following comment https://github.com/openshift/openshift-docs/pull/22882#discussion_r440301564 I used the shorthand. I have now changed it to the longer version as suggested.

kalexand-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -4,48 +4,131 @@
+<6> Across all persistent volume claims in a project, the sum of storage requested in the bronze storage class cannot exceed this value. When this value is set to `0`, the bronze storage class cannot request storage.
+<7> Across all persistent volume claims in a project, the sum of storage requested in the bronze storage class cannot exceed this value. When this value is set to `0`, the bronze storage class cannot create claims.
++
+.. Create a new project request template from the modified `template.yaml` file in the `openshift-config` namespace:
```
~new~

pneedle-rh replied:
Resolved.

kalexand-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -4,48 +4,131 @@
-To require explicit quota to consume a particular resource:
+. Add a resource quota definition to a project request template:
++
+* If a project request template does not exist in a cluster:
```
s/*/**

kalexand-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -4,48 +4,131 @@
++
+By default, the template is called `project-request`.
++
+* If a project request template already exists within a cluster:
```
s/*/**

pneedle-rh replied:
Resolved.

kalexand-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -4,48 +4,131 @@
+$ oc edit template <project_request_template> -n openshift-config
+----
++
+[NOTE]
```
It sounds like you should see this note before you perform the steps, so please move it to the start of the substep.

pneedle-rh replied:
Resolved.

kalexand-rh commented on modules/quotas-requiring-explicit-quota.adoc:
```diff
@@ -4,48 +4,131 @@
+.. Add a resource quota definition, such as the preceding 'storage-consumption' example, into the existing template. The definition must be added before the `parameters:` section in the template.
 
-. Add the following stanza to the master configuration:
+. If you created a new project request template, reference it in the cluster's project configuration resource:
```
~new~

pneedle-rh replied:
Resolved.

kalexand-rh commented:
/cherrypick enterprise-4.2

kalexand-rh commented:
/cherrypick enterprise-4.3

kalexand-rh commented:
/cherrypick enterprise-4.4

kalexand-rh commented:
/cherrypick enterprise-4.5
